module/VuFind/src/VuFind/Search/Solr/CustomFilterListener.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+/**
+ * Solr custom filter listener.
+ *
+ * This can translate a simple filter into a complex set of filters, and it can
+ * "invert" filters by applying Solr filters only when a VuFind filter is absent.
+ *
+ * PHP version 7
+ *
+ * Copyright (C) Villanova University 2022.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Search
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+namespace VuFind\Search\Solr;
+
+use Laminas\EventManager\EventInterface;
+use Laminas\EventManager\SharedEventManagerInterface;
+use VuFindSearch\Backend\BackendInterface;
+
+use VuFindSearch\Service;
+
+/**
+ * Solr custom filter listener.
+ *
+ * @category VuFind
+ * @package  Search
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class CustomFilterListener
+{
+    /**
+     * Backend.
+     *
+     * @var BackendInterface
+     */
+    protected $backend;
+
+    /**
+     * Normal filters
+     *
+     * @var array
+     */
+    protected $normalFilters;
+
+    /**
+     * Inverted filters
+     *
+     * @var array
+     */
+    protected $invertedFilters;
+
+    /**
+     * Name of parameter used to store filters
+     *
+     * @var string
+     */
+    protected $filterParam = 'fq';
```

Comment:
I put this into a property because I think we could theoretically move most of the logic to an abstract base class and then have separate implementations for other search backends like Summon.

Reply:
Ok. This reminds me.. we really should try to move more of the backend-specific stuff to VuFindSearch, I think.

Reply:
I think the original intention of putting the listeners in VuFind instead of VuFindSearch was that the listeners were implementing functionality specific to VuFind rather than generic to backends... but that may no longer be true of all of the listeners, or we may no longer care about that distinction. Certainly happy to revisit. :-)

Reply:
What I meant is that e.g. 'fq' is pretty specific to how the url query syntax with Solr currently works. There's been at least some level of agreement that the legacy facets api that we use should be deprecated with Solr 9, which probably means that in the long run we need to switch to the JSON APIs, for all queries. Then the 'fq' in VuFind classes becomes a burden. So, while I think it's fine that a listener is specific to a backend, I'd like the interface between VuFind and VuFindSearch to be more backend-agnostic. It's for sure not this bug that needs to address it. :)

Reply:
That's a good point, though I don't think there's an easy solution to this problem. The original design decision was to have the contents of the ParamBag passed to the search backend be backend-specific, and to have the Params classes take care of the primary mapping of VuFind parameters to backend parameters. Of course, custom listeners live somewhat outside of this piece, but in the end, I think the problem here is deciding what needs to be given a VuFind-level abstraction, and where we put the mapping logic. Maybe it would be helpful to make facets more abstract so that more code can be shared across backends, but given the inconsistencies in implementations on different platforms (particular with regard to things like AND/OR/NOT facets), that may not be practical. And even if we do decide to make facets uniform, there are going to be other backend-specific features that perhaps do not make sense to develop abstractions for, because the use cases are too narrow. Anyway, you rightly point out that we need to work on the JSON facets implementation, and perhaps that work will inspire our next step forward -- that certainly seems like a good context in which to continue this conversation.